### PR TITLE
feat(providers): strip chain-of-thought markers from custom provider output

### LINKED ISSUE #8590: Chain-of-Thought Leakage Breaking Chat Readability & Poor User Experience
**Please explain the motivation behind the feature request.**
Currently, when using custom LLM providers (via LLMProxy or OpenAI-compatible proxies), Goose displays the model's internal "Chain of Thought" or planning steps (e.g., "Structuring the comparison", "I need to provide a general comparison") as part of the normal chat output. This creates a "Thought Leakage" issue where the UI gets cluttered with internal reasoning, making the actual answer hard to read. It also sometimes causes repetitive loops in responses. This feature would significantly improve usability for users relying on local or proxy-based LLM setups where provider-side filtering is not available.

---

**Describe the solution you'd like**
I would like a configurable setting (toggle or regex-based filter) called **"Sanitize Internal Reasoning"**. This feature should automatically detect and hide or collapse common chain-of-thought patterns (such as lines starting with "Considering", "Structuring", "I will", etc.) before rendering them in the chat UI. If possible, these filtered sections can be shown as a collapsed dropdown (e.g., "Show hidden reasoning") instead of being completely removed.

---

**Describe alternatives you've considered**

* **Prompt Injection (.goosehints):** Attempted to instruct the model to avoid showing reasoning, but this is unreliable and often ignored.
* **Concise Mode:** Enabled concise response style in settings, but it does not prevent reasoning leakage when the model outputs it as plain text.
* **Manual Ignoring:** Currently ignoring these sections manually, which is inefficient and disrupts readability.

---

**Additional context**
This issue primarily affects users using Virtual Keys and local proxy setups (like configure manually provider), where raw model output is passed directly to the UI without any post-processing or filtering.

Example issue:

* The agent repeatedly prints lines like "Structuring the comparison" or "I need to provide a general comparison" before the actual response.

---

* [x] I have verified this does not duplicate an existing feature request

### PR DESCRIPTION
**Pull request by mvanhorn**
## Summary

Custom OpenAI-compatible providers (DeepSeek R1, Qwen-style proxies, LLMProxy, etc.) emit reasoning inline in the content field as `<think>...</think>` or `<thinking>...</thinking>` blocks rather than using the structured `reasoning_content` JSON field. Goose surfaced those blocks directly in chat, cluttering the UI and sometimes triggering repetition.

This change adds a streaming-safe `ThinkFilter` / `split_think_blocks` helper in `crates/goose/src/providers/base.rs` and wires it into both the non-streaming (`response_to_message`) and streaming (`response_to_streaming_message`) paths in `crates/goose/src/providers/formats/openai.rs`. Inline thinking is routed into `MessageContent::thinking`; cleaned content flows through `MessageContent::text`. Structured `reasoning_content` / `reasoning` fields still take precedence, so providers that already send reasoning as a first-class field do not get duplicated output.

Streaming semantics: when a chunk ends on a partial tag prefix (`<thi`, `<think`, `</thinkin`, etc.), the filter holds that suffix until it can decide whether it belongs to a think block or is regular content. A chunk ending mid-tag will never leak characters to the UI that later turn out to be inside a think block.

### Testing

- `cargo fmt --check`: clean
- `cargo clippy -p goose --all-targets -- -D warnings`: clean
- `cargo test -p goose --lib providers::base::tests`: 23 passed (includes new think filter tests)
- `cargo test -p goose --lib providers::formats::openai::tests`: all passed (includes new regression tests)

New unit tests cover:
- Full-content splitter for `<think>`, `<thinking>`, attributes (`<think class="x">`), case-insensitive tags, multiple blocks, plain content, unterminated opens
- Streaming across partial tag boundaries (chunks like `["<thi", "nk>x</thi", "nk>y"]`)
- Non-think tags (`<table>`) are treated as content
- Structured `reasoning_content` path still wins when both are present

### Related Issues
Closes #8590

Prior art: zed's similar fix landed in zed-industries/zed#53488 ("Strip think tags from text stream") for an analogous UI-leakage issue.

This contribution was developed with AI assistance (Codex).

### CODE REVIEW

**Comment by mvanhorn** [comment]
Still resolving routing.

**Comment by mvanhorn** [comment]
Apologies for the noise - I misread my tooling and closed this thinking it was misrouted. block/goose and aaif-goose/goose resolve to the same repo (transfer). Reopening.

**Comment by jh-block** [comment]
Thanks for this @mvanhorn -- can you address the Codex code review feedback (either implement the requested changes or reply to the comments briefly explaining why it's invalid/unnecessary).

**Comment by mvanhorn** [comment]
Thanks @jh-block. Both Codex P2 items addressed in 06d806a:

- parse_think_tag in crates/goose/src/providers/base.rs: added a real-boundary check (>, /, whitespace) after the name match, so `<thinking-mode>`, `<thinking123>`, `<thinker>` no longer get classified as think tags and stripped. New regression tests in base.rs cover that plus `<think data-source="x">` so tags with attributes still work.
- response_to_streaming_message in crates/goose/src/providers/formats/openai.rs: inline `<think>...</think>` chunks now accumulate into pending_inline_thinking and are only flushed at end-of-stream (or at the tool-call flush block). When reasoning_content arrives in any later chunk, saw_structured_reasoning flips and the buffer is cleared, so structured reasoning wins cleanly without duplicated output. New streaming test test_streaming_suppresses_inline_think_when_structured_reasoning_follows covers the race.

Verified: 5 think-filter tests + 58 openai format tests pass, cargo fmt + cargo clippy -D warnings clean.

File: crates/goose/src/providers/formats/openai.rs [hunk 1]
```diff
@@ -935,8 +978,14 @@ where
                 }
 
                 if let Some(text) = text_content {
-                    if !text.is_empty() {
-                        content.push(MessageContent::text(&text));
+                    let filtered = think_filter.push(&text);
+
+                    if !saw_structured_reasoning && !filtered.thinking.is_empty() {
+                        content.push(MessageContent::thinking(filtered.thinking, ""));
+                    }
```
[the remark this answers is not in the document]
Comment (mvanhorn) [reply crates/goose/src/providers/formats/openai.rs:997]: Buffered into `pending_inline_thinking` and only flushed at end-of-stream / tool-call flush. Cleared when `reasoning_content` arrives in a later chunk so structured reasoning wins cleanly. New test `test_streaming_suppresses_inline_think_when_structured_reasoning_follows` covers the race.

File: crates/goose/src/providers/base.rs [hunk 2]
```diff
@@ -24,6 +24,214 @@ use std::pin::Pin;
...
+        if bytes.get(idx) == Some(&b'>') {
+            return Some((ThinkTag::Close, idx + 1));
+        }
+        return None;
+    }
+
+    while let Some(byte) = bytes.get(idx) {
+        if *byte == b'>' {
+            return Some((ThinkTag::Open, idx + 1));
+        }
```
[the remark this answers is not in the document]
Comment (mvanhorn) [reply crates/goose/src/providers/base.rs:228]: Added a real-boundary check (`>`, `/`, whitespace) before returning `ThinkTag::Open`. New tests cover `<thinking-mode>`, `<thinking123>`, `<thinker>`, and `<think data-source="x">` so attributes still work.

File: crates/goose/src/providers/base.rs [hunk 3]
```diff
@@ -24,6 +24,225 @@ use std::pin::Pin;
...
+            return Some((ThinkTag::Close, idx + 1));
+        }
+        return None;
+    }
+
+    // Require a real tag boundary immediately after the name (>, /, or whitespace).
+    // Without this, `<thinking-mode>` or `<thinking123>` would be classified as a
+    // think tag and stripped from normal content.
+    let valid_open_boundary = match bytes.get(idx) {
+        Some(&b) => b == b'>' || b == b'/' || b.is_ascii_whitespace(),
```
[the remark this answers is not in the document]
Comment (jh-block) [reply crates/goose/src/providers/base.rs:218]: I've not seen an empty think tag rendered as a self-closing tag (`<think/>`) before, but here codex is quite confidently saying "for providers that emit self-closing tags" implying that there are some providers that do this. It might be worth being defensive here @mvanhorn?

**Comment by mvanhorn** [comment]
@jh-block fair, defensive handling added in 7940020.

parse_think_tag now distinguishes self-closing (`<think/>`, `<think/>` with attrs, `<thinking/>`, `<think />`) from plain open. I added a `ThinkTag::SelfClosing` variant that is treated as a no-op - the tag is consumed from visible output (so `<think/>` doesn't leak to the user), but `think_depth` and `inside_think` are left unchanged, so the rest of the stream isn't silently swallowed into the thinking bucket.

Regression tests in base.rs cover:
- `<think/>`, `<think />`, `<thinking/>`, `<think data-source="x"/>` -> content preserved, thinking empty
- self-closing inside an open `<think>` block doesn't pre-close it (depth stays at 1 until the real `</think>`)
- streaming case: content after `<think/>` across multiple chunks stays visible

File: crates/goose/src/providers/base.rs [hunk 4]
```diff
@@ -24,6 +24,239 @@ use std::pin::Pin;
...
+        idx += 1;
+    }
+
+    None
+}
+
+fn is_possible_partial_think_tag(suffix: &str) -> bool {
+    static OPEN_RE: LazyLock<Regex> = LazyLock::new(|| {
+        Regex::new(r"(?is)^<(?:t(?:h(?:i(?:n(?:k(?:i(?:n(?:g)?)?)?)?)?)?)?)(?:\s[^>]*)?$").unwrap()
+    });
```
[the remark this answers is not in the document]
Comment (jh-block) [reply crates/goose/src/providers/base.rs:272]: @mvanhorn looks like there's one last issue to resolve here (the streaming parser can emit `<think/` or `<thinking/` if that appears at the end of a chunk). If you can tweak the regex to handle that I think then we should be able to merge this

**Comment by mvanhorn** [comment]
@jh-block addressed in bb9f824. Extended OPEN_RE with a trailing `/` alternative so `<think/` and `<thinking/` at a chunk boundary buffer until the `>` arrives in the next chunk.

Root cause: the old regex accepted `<think<whitespace>...` but not a bare trailing `/`, so a chunk ending mid self-closing tag fell out of `is_possible_partial_think_tag` and leaked the `<think/` prefix into visible content.

Added `test_think_filter_streaming_across_self_closing_boundary` covering three partial-boundary cases: `<think/` + `>`, `<thinking/` + `>`, and `<think ` + `/>`. Asserts no `<` leaks into content and thinking stays empty (self-closing = no payload). Verified with `cargo test -p goose --lib providers::base::tests` - 29 tests pass, clippy/fmt clean.

**Comment by mvanhorn** [comment]
@jh-block addressed in ee82ace. The open-tag parser and partial-tag path now track quoted attribute values via a new `contains_unquoted_gt` helper, so `<think data="a>b"/>Visible` (and the single-quote variant) is treated as a self-closing tag with `Visible` flowing through cleanly. Tests cover both quote chars in self-closing position, the open/close-with-attrs case, and a streaming chunk boundary inside the quoted value. cargo fmt/clippy/test all green locally.

**Review by jh-block — APPROVED** [review]
